nuget/lib/dependabot/nuget/file_updater/project_file_declaration_finder.rb

```diff
@@ -95,6 +111,80 @@ def declaring_file
 
           raise "No file found with name #{filename}!"
         end
+
+        def sdk_import_strings
+          sdk_strings(SDK_IMPORT_REGEX, "Import", "Sdk", "Version")
+        end
+
+        # rubocop:disable Metrics/PerceivedComplexity
+        def sdk_project_strings
+          dep_name = dependency_name&.downcase
+          dep_version = declaring_requirement.fetch(:requirement)
+          strings = []
+          declaring_file.content.scan(SDK_PROJECT_REGEX).each do |string|
+            xml = string
+            xml += "</Project>" unless string.end_with?("/>")
+            node = Nokogiri::XML(xml)
+            node.remove_namespaces!
+            element = node.at_xpath("/Project")
+            next unless element
+
+            sdk_references = element.attribute("Sdk")&.value ||
+                             element.attribute("sdk")&.value
+            sdk_references = sdk_references&.strip
+            next unless sdk_references&.include?("/")
+
+            sdk_references.split(";").each do |sdk_reference|
+              parts = sdk_reference.split("/")
+              next unless parts.length == 2
+              next unless parts[0]&.downcase == dep_name
+              next unless parts[1] == dep_version
+
+              strings << sdk_reference
+            end
+          end
+          # Drop duplicates, to avoid potential issues with repeated replacements (e.g. upgrading Foo/1.0.0-beta to
+          # Foo/1.0.0-beta.2). However, that turns the empty array into nil.
+          strings.uniq! || []
+        end
+        # rubocop:enable Metrics/PerceivedComplexity
+
+        def sdk_sdk_strings
+          sdk_strings(SDK_SDK_REGEX, "Sdk", "Name", "Version")
+        end
+
+        # rubocop:disable Metrics/CyclomaticComplexity
+        # rubocop:disable Metrics/PerceivedComplexity
+        def sdk_strings(regex, element_name, name_attribute, version_attribute)
+          dep_name = dependency_name&.downcase
+          dep_version = declaring_requirement.fetch(:requirement)
+          strings = []
+          declaring_file.content.scan(regex).each do |string|
+            xml = string
```

Comment:
```suggestion
          declaring_file.content.scan(regex).each do |xml|
```

Reply:
Applied

Reply:
Actually, I'm having to revert this - `string` needs to be used unchanged afterwards, while `xml` potentially gets modified.

Reply:
fwiw string will also get modified:

```ruby
irb(main):004:0> string = "foo"
irb(main):005:0> xml = string
irb(main):006:0> xml << "x"
=> "foox"
irb(main):007:0> string
=> "foox"
```

Reply:
It would if I used `<<`, but I use `xml +=` and it's my understanding that that creates a new string.

Reply:
Ah that's correct, but looking at the code we never refer to `string` after `xml` is changed, at least not now that the code is extracted to `parse_element`. Anyway, it's not really important I guess

Reply:
We do in `sdk_strings`: `strings << string`; that's why parse_element() retains the `+=`.